filterExprParser.go

```diff
@@ -0,0 +1,1265 @@
+	}
+}
+
+func (arg *FEConstFuncTwoArgsName) OutputExpression() (string, error) {
+	if arg.Atan2 != nil && *arg.Atan2 == true {
+		return MathFuncAtan2, nil
+	} else if arg.Power != nil && *arg.Power == true {
+		return MathFuncPow, nil
+	} else {
+		return "?? (FEConstFuncTwoArgsName)", ErrorNotFound
+	}
+}
+
+type FEBooleanFuncExpr struct {
+	BooleanFuncTwoArgs *FEBooleanFuncTwoArgs `@@`
+}
+
+func (e *FEBooleanFuncExpr) String() string {
+	if e.BooleanFuncTwoArgs != nil {
+		return e.BooleanFuncTwoArgs.String()
+	} else {
+		return "?? (FEBooleanFuncExpr)"
+	}
+}
+
+func (f *FEBooleanFuncExpr) OutputExpression() (Expression, error) {
+	if f.BooleanFuncTwoArgs == nil {
+		return nil, fmt.Errorf("Invalid FEBooleanFuncExpr")
+	}
+	return f.BooleanFuncTwoArgs.OutputExpression()
+}
+
+type FEBooleanFuncTwoArgs struct {
+	BooleanFuncTwoArgsName *FEBooleanFuncTwoArgsName `( @@ "("`
+	Argument0              *FEConstFuncArgument      `@@ ","`
+	Argument1              *FEConstFuncArgumentRHS   `@@ ")" )`
+}
+
+func (a *FEBooleanFuncTwoArgs) String() string {
+	if a.BooleanFuncTwoArgsName != nil {
+		return fmt.Sprintf("%v( %v , %v )", a.BooleanFuncTwoArgsName.String(), a.Argument0.String(), a.Argument1.String())
+	} else {
+		return "?? (FEBooleanFuncTwoArgs)"
+	}
+}
+
+func (f *FEBooleanFuncTwoArgs) OutputExpression() (Expression, error) {
+	if f.BooleanFuncTwoArgsName != nil && f.BooleanFuncTwoArgsName.RegexContains != nil && *f.BooleanFuncTwoArgsName.RegexContains {
+		outputExpr, err := f.BooleanFuncTwoArgsName.OutputExpression()
+		if err != nil {
+			return nil, err
+		}
+		outExpr := outputExpr.(LikeExpr)
+
+		arg0, err := f.Argument0.OutputExpression()
+		if err != nil {
+			return outExpr, err
+		}
+		outExpr.Lhs = arg0
+
+		arg1, err := f.Argument1.OutputRegexExpression()
+		if err != nil {
+			return outExpr, err
+		}
+		outExpr.Rhs = arg1
+
+		return outExpr, nil
+	} else {
+		return nil, fmt.Errorf("Invalid FEBooleanFuncTwoArgs %v", f.BooleanFuncTwoArgsName.String())
+	}
+}
+
+type FEBooleanFuncTwoArgsName struct {
+	RegexContains *bool `@"REGEXP_CONTAINS"`
+}
+
+func (n *FEBooleanFuncTwoArgsName) String() string {
+	if n.RegexContains != nil && *n.RegexContains == true {
+		return FuncRegexp
+	} else {
+		return "?? (FEBooleanFuncTwoArgsName)"
+	}
+}
+
+func (n *FEBooleanFuncTwoArgsName) OutputExpression() (Expression, error) {
+	if n.RegexContains != nil && *n.RegexContains == true {
+		return LikeExpr{}, nil
+	} else {
+		return nil, ErrorNotFound
+	}
+}
+
+func parserWrapper(parser *participle.Parser, expression string, fe *FilterExpression, err *error) {
+	defer func() {
+		if r := recover(); r != nil {
+			*err = fmt.Errorf("Error from parser: %v", r)
+		}
+	}()
+
+	*err = parser.ParseString(expression, fe)
+}
+
+func NewFilterExpressionParser(expression string) (*participle.Parser, *FilterExpression, error) {
+	fe := &FilterExpression{}
+	parser, err := participle.Build(fe)
+	if err != nil || len(expression) == 0 {
```

Comment:
when len(expr) =0, don't you need to create a not-nil err?

Reply:
Yes. Created.